site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,382 @@
+
+Let's set `client_id`, `client_secret`, `idp_url` (identity provider url), and `redirect_uri` values using `development.rb`.
+I chose to set these values here as they will likely change in staging or production environments, but feel free to set them based on your own standards.
+We will also initialize an environment variable that contains our "HMAC secret" that will be used to decode the token received in the authentication process.
+We want to keep this as secure as possible and therefore we will set it at run-time.
+
+```ruby
+  # OAuth configuration
+  config.x.oauth.client_id = "my-client-id"
+  config.x.oauth.client_secret = "my-super-secret-oauth-secret"
+  config.x.oauth.idp_url = "http://localhost:9011/"
+  config.x.oauth.redirect_uri = "http://localhost:3000/oauth2-callback"
+  config.x.oauth.hmac = ENV['HMAC_SECRET']
+```
+
+### Authenticating with FusionAuth
+Before we get into the code, let's take a look at the Authentication Code grant flow as it applies to our Rails app, the user, and FusionAuth:
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/rails-fa-sequence.png" alt="A sequence diagram illustrating the authentication code grant flow." class="img-fluid" figure=false %}
+ 
+This gives insight into the routes that we will need to define in our Rails app. 
+
+### Routes
+For this app, we are going to setup four routes:
+* **Welcome page**: Our root page. 
+* **Login endpoint**: Makes a request for the FusionAuth login page when a user clicks on the login link. 
+* **OAuth callback endpoint**: Receiver of the redirect from FusionAuth when the authorization code is granted.
+* **Logout endpoint**: Receiver of the redirect from FusionAuth when successfully logging out.
+
+ ```ruby
+ Rails.application.routes.draw do
+   root to: 'welcome#index'
+   get '/oauth2-callback', to: 'oauth#oauth_callback'
+   get '/logout', to: 'oauth#logout'
+   get '/login', to: 'oauth#login'
+ end
+ ```
+
+#### The Welcome page
+Let's start off with a very basic welcome view that contains a link for a user to login to the application.
+This page will serve as the root page for our application.
+
+```ruby
+# app/views/welcome/index.html.erb
+
+<p>Welcome! Please <%= link_to 'sign in', controller: 'oauth', action: :login %> </p>
+```
+
+It should look something like this when rendered.
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/home.png" alt="The example Home screen for the Rails application." class="img-fluid" figure=false %}
+
+#### OauthController
+
+The "OauthController" will handle each of the remaining routes needed to complete the authorization code grant flow.
+First, we will setup our client using the [oauth2](https://github.com/oauth-xx/oauth2) gem. This provides us with a "REST" client wrapper for the OAuth 2.0 specification.
```

Comment:
I don't understand what you mean by "REST"?

Reply:
Added a [link](https://www.codecademy.com/articles/what-is-rest) :)